snmp/datadog_checks/snmp/models.py

```diff
@@ -23,46 +24,29 @@ class OID(object):
     def __init__(self, value):
         # type: (Union[Sequence[int], str, ObjectName, ObjectIdentity, ObjectType]) -> None
         parts = None  # type: Optional[Tuple[int, ...]]
+        object_identity = None  # type: Optional[ObjectIdentity]
 
         try:
             parts = parse_as_oid_tuple(value)
         except CouldNotDecodeOID:
             raise  # Invalid input.
         except UnresolvedOID:
+            # `value` must be a PySNMP object, let's inspect it.
             if isinstance(value, ObjectType):
-                # An unresolved `ObjectType(ObjectIdentity('<MIB>', '<symbol>'))`.
-                parts = None
+                object_identity = object_identity_from_object_type(value)
             elif isinstance(value, ObjectIdentity):
-                # An unresolved `ObjectIdentity('<MIB>', '<symbol>')`.
-                parts = None
+                object_identity = value
             else:  # pragma: no cover
                 raise RuntimeError('Unexpectedly treated {!r} as an unresolved OID'.format(value))
-
-        # Resolve the `ObjectIdentity` which we can use to resolve the MIB name of the OID (for metric naming).
-        # PySNMP objects may contain MIB information already, so check for them in priority.
-        if isinstance(value, ObjectType):
-            # No other choice than to use private API here.
-            object_identity = value._ObjectType__args[0]
-            if not isinstance(object_identity, ObjectIdentity):  # pragma: no cover
-                raise RuntimeError('Expected {!r} to be an `ObjectIdentity` instance'.format(object_identity))
-        elif isinstance(value, ObjectIdentity):
-            object_identity = value
+            parts = parts_from_object_identity(object_identity)
         else:
-            # Fallback.
-            if parts is None:  # pragma: no cover
-                raise RuntimeError('`parts` should have been set')
             object_identity = ObjectIdentity(parts)
 
         self._parts = parts
-        self._object_identity = object_identity  # type: ObjectIdentity
+        self._object_identity = object_identity
 
     def resolve(self, mib_view_controller):
         # type: (MibViewController) -> None
-        if self._parts is not None:
-            # Client code should only call this if they're certain the
-            # underlying OID isn't resolved yet.
-            raise RuntimeError('Already resolved as {}'.format(self._parts))
```

Comment:
Removed this check because it's not actually necessary - it's safe to resolve a MIB symbol multiple times, and may be required even if we already have the `parts` (eg to later call `.get_mib_symbol()`.

Reply:
Where is `OID.resolve` used ? other than this case in test https://github.com/DataDog/integrations-core/blob/5813bf017d0fb5b124be5f8ac5379180ebdaa417/snmp/tests/test_utils.py#L109

Reply:
Hmm, you're right, it's not used yet. I added it in https://github.com/DataDog/integrations-core/pull/6242 and meant to push a follow-up PR for using this in `OIDResolver`, but haven't done it yet.